adr/ADR-40.md

```diff
@@ -0,0 +1,72 @@
+# Request Many
+
+| Metadata | Value                 |
+|----------|-----------------------|
+| Date     | 2023-06-26            |
+| Author   | @aricart, @scottf     |
+| Status   | Partially Implemented |
+| Tags     | client                |
+
+## Problem Statement
+Have the client support receiving multiple replies from a single request, instead of limiting the client to the first reply.
+
+## Design
+
+Making the request and handling multiple replies is straightforward. Much like a simplified fetch, the developer 
+will provide some basic strategy information that tell how many messages to wait for and how long to wait for those messages.
+
+* The client doesn't assume success or failure - only that it might receive messages.
+The various options are there to short circuit the length of the wait.
+* The Subsequent Max value allows for waiting for the service with the slowest latency. (scatter gather)
+* The message count allows for waiting for some count of messages or a timer (scatter gather)
+
+## Options
+
+#### Max Wait
+
+The maximum amount of time to wait for responses.
+
+* Responses are accepted until the max time is reached.
+
+#### First Max / Subsequent Max
+
+A combination time...
+
+* The first response must be received by the first max time or the request terminates.
+* Subsequent messages must be received in the gap time, the time since the last message was received or the request expires.
+* Each time a message is received the gap timer is reset.
+
+#### Count
+
+The number of responses to allow.
+
+* Responses are accepted until the count is reached.
+* Must be used in combination with a time as the count may never be reached.
+
+#### Sentinel
```

Comment:
As discussed on calls should we be clear that this is not something to add to the clients but rather a pattern to document for now?

like, this isnt hard and its more flexible:

```go
MultiSub(subj, func(m *nats.Msg)) {
  // sentinal that can be based on nil body, headers or anything
  if len(m.Data)==0 || m.Header.Get("EOF") != "" {
    m.Sub.Unsubscribe()
    return
  }

  // handle msg
}
```

Reply:
It depends, right - if you are yielding an iterator, you could break, and that would close it. So for example if you are using the mux subscription to handle the responses, you cannot unsubscribe there,  and perhaps you want to cleanup there. Not sure that not implementing it is not helpful.

Imho, if it is a pattern, the question is whether the sentinel has any meaning, if it doesn't, the client shouldn't even see it.

Reply:
The concern is the sentinel can be anything - server supports nil in one specific case but users might other things. 

So to have it built in the se final detector should be an injectable dependency so we provide a nil detecting one and users can do their own detector?

Reply:
As usual, maybe lets remove it then. We can always extend it later.